### ISSUE TITLE
Write columns in parallel

### ISSUE THREAD

**Issue by derekperkins** [parquet-go/parquet-go#123]
Our bottleneck in writing ~500MB zstd compressed files is that the write happens single threaded, often taking minutes to complete. It also takes ~15 GB of RAM during that processing, which is a weird ratio for cloud resources. Writing / compressing in parallel would significantly increase our throughput.

There's prior art in the rust implementation, which hopefully translates well to Go.
- https://github.com/apache/arrow-rs/pull/4871

**Comment by achille-roussel** [comment]
Hello @derekperkins 

Would you happen to have a snippet of code that's representative of the workload and we could use to benchmark? If not, maybe you could share CPU and memory profiles.

Generating the file in parallel can be done, with some limits, I have doubts that it be effective to parallelize the generation of a single row group; I see how generating multiple row groups in parallel could be effective, this is partially supported today since you could generate multiple files with one row group each, then merge them into one (much easier to optimize the copy of file sections).

15GB of memory used to generate a 500MB file seems high tho, parquet-go was designed to have a very low memory footprint, so I'm thinking something else might be going on that results in the poor performance here.

**Comment by derekperkins** [comment]
I posted a few months ago about how I wasn't getting the same compression when writing a row at a time, so I'm now reading in 5k ~1MB HTML files, then making a single call to write with all the rows already fully populated.
- https://github.com/parquet-go/parquet-go/issues/89

### Allocated Heap Profile

As expected, I'm allocating the majority of memory reading in my raw html bytes, with parquet-go only allocating ~11% of the total.

<img width="890" alt="image" src="https://github.com/parquet-go/parquet-go/assets/3588778/37ecbab6-a6d1-4ee8-88bb-a812f98a1999">

### CPU Profile

Similarly, CPU time is dominated by zstd compression on the write side, and gzip decompression of the read side.

<img width="1020" alt="image" src="https://github.com/parquet-go/parquet-go/assets/3588778/010bb892-a429-453c-9d5f-7bb2a7379500">

**Comment by achille-roussel** [comment]
Thanks for the extra info.

Just to have a full picture, do you know which zstd settings you're using? and which CPU architecture you're running on?

I'm also curious about configuration options in the writer like the page buffer size, which could also have an impact on writer performance.

**Comment by derekperkins** [comment]
This is running on amd64 on Google Cloud.

I'm not setting much, so I'm very open to the idea that I'm not using the library in the best way.

```go
w := parquet.NewGenericWriter[parquetRow](gcsWriter,
    parquet.SchemaOf(&parquetRow{}),
    parquet.Compression(&zstd.Codec{Level: zstd.SpeedBestCompression}),
)
```

**Comment by achille-roussel** [comment]
Alright, thanks for the info!

I assume you want the maximum compression ratio, which is why you configure it explicitly, which is likely why your program is spending so much time in the zstd encoder (you're telling it to use as much CPU as possible to get the best compression).

Something I would try, is adding a configuration parameter to the zstd codec that enables parallel compression (see https://pkg.go.dev/github.com/parquet-go/parquet-go@v0.20.1/compress/zstd#Codec, and https://github.com/parquet-go/parquet-go/blob/v0.20.1/compress/zstd/zstd.go#L56).

Since most of the time is spent in that part of the code, it might allow you to leverage multiple cores to speed up this part of the processing without having to make changes to the parquet API itself.

Do you want to give it a try?

**Comment by derekperkins** [comment]
Yeah, that sounds great

**Comment by derekperkins** [comment]
PR submitted adding that option

**Comment by derekperkins** [comment]
Did my first test with concurrency set to 4, and it didn't seem to have any impact. The charts below depict three successive runs of the same file at different compression levels, Best, Better, and Default, in that order.

| Compression | Concurrency | Time   | Filesize |
|-------------|-------------|--------|----------|
| Best        | 1           | 153.0s  | 425 MB   |
| Best        | 4           | 149.5s | 425 MB   |
| Better      | 4           | 56.4s  | 451 MB   |
| Default     | 4           | 19.9s  | 554 MB   |


### CPU

<img width="839" alt="image" src="https://github.com/parquet-go/parquet-go/assets/3588778/dd585beb-e440-49cc-b131-92951f5aa523">


### RAM

<img width="842" alt="image" src="https://github.com/parquet-go/parquet-go/assets/3588778/9e923d37-a92f-47f1-aaf3-c1c7285199a8">

**Comment by achille-roussel** [comment]
Reopening to continue the discussion.

I apologize, I misinterpreted the purpose of `WithEncoderConcurrency`; looking into the implementation, this seems to allow concurrent calls to `EncodeAll` in the zstd encoder, but each call still processes data in a single goroutine. Digging more into the implementation, it doesn't look like there is a way to parallelize compression. To achieve this, we would need to perform compression asynchronously internally in the parquet writer, so pages can be compressed in parallel, that's a significant rewrite of the internal tho.

Another setting that you could be working with is the page buffer size https://pkg.go.dev/github.com/parquet-go/parquet-go#PageBufferSize, say you increased it to 1 MiB, would that improve the compression throughput? The reason I'm suggesting it is because it looks like on the CPU profiles, a lot of time is spent in the intermediary frames within `EncodeAll` (e.g., `hashLen`), these would get amortized if your program had less pages to compress.

Once again, I imagine the ~20% better compression really matters to your use case? Otherwise there's an easy ~10x improvements using the default zstd compression settings it seems :)

**Comment by achille-roussel** [comment]
🤔 digging more into this, it might not work either because the implementation of the zstd encoder has a hardcoded block size to 128KiB https://github.com/klauspost/compress/blob/master/zstd/blockdec.go#L43-L44, since the default page size is larger than that already.

**Comment by derekperkins** [comment]
> Once again, I imagine the ~20% better compression really matters to your use case

Yes, this project is all about cost savings in long term storage.

I very much appreciate your help thinking through this. It seems like I may just be bumping up against some hard limits. It seems like probably the only thing left to try is to reduce RAM consumption. There's [this reference](https://pkg.go.dev/github.com/parquet-go/parquet-go#readme-c-using-on-disk-page-buffers) to using on-disk page buffers in the docs, but given that the majority of RAM consumption is from my in-memory representation, I'm not sure how much that's going to save. 

In my [other issue](https://github.com/parquet-go/parquet-go/issues/89), it seems like there is a loss of compression when writing a subset of the data. To be fair, I only tested the two extremes, 1 single row at a time and all 5000 rows together. Maybe there's a middle ground where I can fetch chunks of rows and maybe do 5 calls to `w.Write` with 1000 rows each, which should reduce my memory usage 70-80% while still getting the compression benefits. I'll probably just need to test it out.

**Comment by gnagel** [comment]
Following this one here, my issues are nearly identical - it's taking upto 1hr to write a block of 5Gib JSON data.

I'm going to test out the suggestion here of tuning the batch size and writer concurrency.

![Screenshot 2024-04-19 at 6 39 36 AM](https://github.com/parquet-go/parquet-go/assets/619921/36518218-14be-42da-8e26-f19bea84dca4)

**Comment by achille-roussel** [comment]
Hello @gnagel 

Based on the profile you shared, it looks like your case might be different. Only a subfraction of your time is spent compressing (the orange area on the left), the rest seems to be sent in the parquet-go library.

Would you be able to share the schema you're using?

**Comment by derekperkins** [comment]
@lukealonso yes, very interested

**Comment by derekperkins** [comment]
Just now getting back to this. Splitting up row group writing didn't have an appreciable effect on RAM consumption or CPU (both of these written with zstd fastest), but it was significantly slower with 4 row groups, and slightly larger data.

### 4 Row Groups of 1,250

Latency: 76s
Filesize: 544 MB

<img width="1269" alt="image" src="https://github.com/parquet-go/parquet-go/assets/3588778/55ea906a-3b90-40ab-9dca-1c6e66fac041">

### 1 Row Group of 5,000

Latency: 58s
Filesize: 536 MB

<img width="1289" alt="image" src="https://github.com/parquet-go/parquet-go/assets/3588778/dead8cab-55ab-40b0-85a2-acb283bb27ad">

@lukealonso any chance you can share your patch?

**Comment by yusufozturk** [comment]
Just checking if there is any progress on this @lukealonso @derekperkins 
Thanks.

**Comment by rockwotj** [comment]
BTW it looks like the patches are in @lukealonso's public repo: https://github.com/lukealonso/parquet-go/blob/213b4d466b8eff49424c95b3983994ffab69c6ac/writer.go#L432-L444

**Comment by rockwotj** [comment]
@lukealonso any interest in upstreaming that? @achille-roussel would parquet-go be OK with adopting an API like that? I could probably find time to contribute a similarish API

**Comment by achille-roussel** [comment]
Always happy to add new APIs if they solve real user problems!

The concurrent row group writer is a very interesting model, let's get this merged in @lukealonso ?

**Comment by rockwotj** [comment]
I am planning on doing some benchmarking next week to see if parallel parquet construction helps out my use case and will report back
